### DISCUSSION TITLE
Issue with KMK only outputting 1 key

### DISCUSSION THREAD

**Discussion by dlip** [manna-harbour/xmk#7] [Q&A]
I have a mapping for my laptop
```
q w e r t y u i o p 
a s d f g h j k l semicolon 
z x c v b n m comma dot slash 
leftalt space rightalt 
```
and a keymap
```
keyboard.keymap = [
    [KC.Q],    [KC.W],    [KC.F],    [KC.P],    [KC.B],             [KC.J],    [KC.L],    [KC.U],    [KC.Y],    [KC.QUOT],
    [KC.A],    [KC.R],    [KC.S],    [KC.T],    [KC.G],             [KC.M],    [KC.N],    [KC.E],    [KC.I],    [KC.O],
    [KC.Z],    [KC.X],    [KC.C],    [KC.D],    [KC.V],             [KC.K],    [KC.H],    [KC.COMM], [KC.DOT],  [KC.SLSH],
                                                [KC.ESC],  [KC.SPC],[KC.TAB],
]
```
and in `xmk_scanner.py` i've added to line 20
```
print(key, press)
```

then when pressing keys i get successful output:
```
0 True
0 False
1 True
1 False
2 True
2 False
3 True
3 False
...
```
but for some reason only the first key (id 0) works to output 'q'
if i change the next line from
```
            return keypad.Event(key, press)
```
to

```
            return keypad.Event(0, press)
```
any key outputs 'q' as expected
but if i change it to 1 or above, nothing gets output
```
            return keypad.Event(1, press)
```
The only thing i can think of is something changed in circuit python, i think you must be using v7 since i had to comment this line out in `boot.py` that doesn't exist in v8 anymore
```
# supervisor.set_next_stack_limit(4096 + 4096)
```

do you have any other ideas?

**Comment by dlip** [comment]
printing `supervisor.runtime.next_stack_limit` = `24596`

**Comment by manna-harbour** [comment]
Did you try an older circuit python? If you're using my KMK branch directly then I imagine it's a circuit python issue. I'm not sure offhand what version I used but it was the current version at the time of writing the xmk keyboard definition.

**Reply by dlip** [reply]
I reverted to circuit python v7.33 and reset all the code to what is in your branch, but i got the same result: the keys register in the scanner but only the 'q' key outputs anything... is there anything else to try?

**Reply by manna-harbour** [reply]
What MCU are you using? Have you tried the default keymap?

**Reply by dlip** [reply]
Its the WaveShare RP2040-Zero. Yes, I used your default keymap

**Reply by manna-harbour** [reply]
Have you tried the MCU with another keyboard definition (in a keyboard, or jumping pins on the bare MCU)?

**Reply by manna-harbour** [reply]
Is there any error output? Have you tried enabling debugging?
https://github.com/KMKfw/kmk_firmware/blob/master/docs/en/debugging.md

**Reply by dlip** [reply]
I haven't used it in a keyboard, i just had a spare one laying around i thought i'd play with. I thought the xmk scanner would just send keycodes to the usb directly?

With debugging enabled, there is some hint in the log about whats going on:

`MatrixUndefinedCoordinate(ic=1)`

```
0 True
14212910 kmk.kmk_keyboard: MatrixChange(ic=0, pressed=True)
14212916 kmk.keys: TRNS: Key(code=1000, has_modifiers=None)
14212921 kmk.kmk_keyboard: KeyResolution(key=Key(code=20, has_modifiers=None))
14212929 kmk.kmk_keyboard: coordkeys_pressed={0: Key(code=20, has_modifiers=None)}
14212934 kmk.kmk_keyboard: keys_pressed={Key(code=20, has_modifiers=None)}
0 False
14213251 kmk.kmk_keyboard: MatrixChange(ic=0, pressed=False)
14213255 kmk.kmk_keyboard: KeyResolution(key=Key(code=20, has_modifiers=None))
14213261 kmk.kmk_keyboard: coordkeys_pressed={}
14213265 kmk.kmk_keyboard: keys_pressed=set()
1 True
14221672 kmk.kmk_keyboard: MatrixChange(ic=1, pressed=True)
14221676 kmk.kmk_keyboard: KeymapIndexError(idx=1, layer=0)
14221679 kmk.kmk_keyboard: MatrixUndefinedCoordinate(ic=1)
14221683 kmk.kmk_keyboard: coordkeys_pressed={}
14221687 kmk.kmk_keyboard: keys_pressed=set()
1 False
14221888 kmk.kmk_keyboard: MatrixChange(ic=1, pressed=False)
14221892 kmk.kmk_keyboard: KeyNotPressed(ic=1)
14221896 kmk.kmk_keyboard: KeymapIndexError(idx=1, layer=0)
14221899 kmk.kmk_keyboard: MatrixUndefinedCoordinate(ic=1)
14221903 kmk.kmk_keyboard: coordkeys_pressed={}
14221907 kmk.kmk_keyboard: keys_pressed=set()
```
